cmake/cmaize/toolchain/toolchain.cmake

```diff
@@ -60,38 +56,80 @@ cpp_class(Toolchain)
         endif()
 
-        Toolchain(SET "${self}" file_path "${toolchain_path}")
-
         # Initialize the toolchain object
         Toolchain(__initialize "${self}")
 
-        # Parse the user toolchain into a list of lines
+        # Read the toolchain file
         file(READ "${toolchain_path}" file_contents)
-        # string(REPLACE "\"" "\\\"" file_contents "${file_contents}")
-        
-        # Parse user options
-        Toolchain(_parse_toolchain "${self}" "${file_contents}")
+
+        # Store toolchain contents
+        cpp_encode_special_chars("${file_contents}" encoded_file_contents)
+        Toolchain(SET "${self}" encoded_toolchain_contents "${encoded_file_contents}")
 
     endfunction()
 
     #[[[
     # Constructor for Toolchain object with a string assumed to be the
     # contents of an already opened toolchain file.
     #
-    # :param toolchain_contents: Contents of the toolchain file.
-    # :type toolchain_contents: str
+    # :param file_contents: Contents of the toolchain file.
+    # :type file_contents: str
     #]]
     cpp_constructor(CTOR Toolchain str)
-    function("${CTOR}" self toolchain_contents)
-
-        # No file provided
-        Toolchain(SET "${self}" file_path "")
+    function("${CTOR}" self file_contents)
 
         # Initialize the toolchain object
         Toolchain(__initialize "${self}")
 
-        # Parse user options
-        Toolchain(_parse_toolchain "${self}" "${toolchain_contents}")
+        # Store toolchain contents
+        cpp_encode_special_chars("${file_contents}" encoded_file_contents)
+        Toolchain(SET "${self}" encoded_toolchain_contents "${encoded_file_contents}")
+
+    endfunction()
+
+    cpp_member(generate_file_contents Toolchain str)
+    function("${generate_file_contents}" self return_value)
+
+        # Initialize the file contents with a warning
+        set(file_string
+            "# This file was generated by CMaize. Any changes made to this
+# file will be overwritten the next time it is generated.\n\n"
+        )
+
+        # Get the autopopulated options
+        Toolchain(GET "${self}" auto_opt_map auto_options)
+        cpp_map(KEYS "${auto_opt_map}" keys)
+
+        # Start the autopopulated section
+        string(APPEND file_string "# CMaize Autopopulated Options\n\n")
+
+        # Add the autopopulated options
+        foreach(key ${keys})
+            cpp_map(GET "${auto_opt_map}" value "${key}")
+            string(APPEND file_string "set(${key} ${value})\n")
+        endforeach()
+
+        # Start the user toolchain section
+        string(APPEND file_string "\n# User Toolchain\n\n")
+
+        # Add the user toolchain
+        Toolchain(GET "${self}" file_contents encoded_toolchain_contents)
+        string(APPEND file_string "${file_contents}")
+
+        cpp_decode_special_chars("${file_string}" file_string)
+
+        set("${return_value}" "${file_string}" PARENT_SCOPE)
```

Comment:
FWIW I actually made a [`return`](https://github.com/CMakePP/CMakePPLang/blob/master/cmake/cmakepp_lang/utilities/return.cmake) function. Presently it's the same as what you have, but in more complicated cases it'll also return scope to the caller.

Reply:
I'll give that a shot. I was worried about what it might do to the escape characters here, but it will be a good test!

Reply:
I think the escapes will be fine because it actually takes the name of the variable to return (in your case it would be `cpp_return(return_value)`, so line 119 would need to decode into `return_value` and not `file_string`).

Reply:
The `test_generate_file_contents` test fails when I change it to:
```cmake
cpp_decode_special_chars("${file_string}" return_value)

cpp_return(return_value)
```

It looks like the return variable comes back as an empty string. Any suggestions?

Reply:
Oh wait, I misused my own function. Let me try that again.

Reply:
Try:
```
cpp_decode_special_chars("${file_string}" "${return_value}") cpp_return("${{return_value}")
```
The value you're trying to set in the parent scope is not `return_value` as I originally had it, but whatever the value of `return_value` is.

Reply:
Got it, that worked. Thank you!